File: python/etudiant.py
"""Module Etudiant — Hérite de Personne.

TP1 : héritage (super().__init__) + ajoute numero_etudiant, moyenne, liste_cours.
TP2 : encapsulation moyenne + numero_etudiant en lecture seule + liste_cours protégée.
TP3 : afficher_details() override polymorphe.
TP4 : Pattern Observer — Etudiant est le *Sujet Observable*.
      Quand la moyenne est modifiée, tous les observateurs enregistrés sont notifiés.

SRP : Etudiant représente l'identité académique d'un étudiant.
LSP : substituable à Personne — afficher_details() retourne une str, ne lève rien.
OCP : le mécanisme d'observation est ouvert à l'extension (n'importe quel Observer
      peut s'abonner) sans modifier la classe.
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from typing import List

from cours import Cours
from personne import Personne


# ---------------------------------------------------------------------------
# Pattern OBSERVER — interface Observer (DIP : dépendance vers l'abstraction)
# ---------------------------------------------------------------------------
class Observer(ABC):
    """Interface observateur : reçoit les mises à jour d'un Etudiant."""

    @abstractmethod
    def mise_a_jour(self, etudiant: "Etudiant") -> None:
        """Appelée automatiquement quand la moyenne d'un étudiant change."""


# ---------------------------------------------------------------------------
# Classe Etudiant (Sujet Observable)
# ---------------------------------------------------------------------------
class Etudiant(Personne):
    """Un étudiant : une Personne avec un numéro, une moyenne et des cours."""

    MOYENNE_MIN = 0.0
    MOYENNE_MAX = 20.0

    def __init__(
        self,
        nom: str,
        age: int,
        numero_etudiant: str,
        moyenne: float = 0.0,
    ):
        super().__init__(nom, age)
        if not numero_etudiant or not str(numero_etudiant).strip():
            raise ValueError("Le numéro d'étudiant ne peut pas être vide.")
        self.__numero_etudiant = str(numero_etudiant).strip()
        self.__liste_cours: List[Cours] = []
        self.__observers: List[Observer] = []
        # passage par le setter pour validation (ne notifie pas encore car
        # _notifier vérifie que des observers existent)
        self.moyenne = moyenne

    # ----- Observer : abonnement / désabonnement -----
    def ajouter_observer(self, obs: Observer) -> None:
        if obs not in self.__observers:
            self.__observers.append(obs)

    def retirer_observer(self, obs: Observer) -> None:
        self.__observers.remove(obs)

    def _notifier(self) -> None:
        for obs in self.__observers:
            obs.mise_a_jour(self)

    # ----- numero_etudiant : lecture seule -----
    @property
    def numero_etudiant(self) -> str:
        return self.__numero_etudiant

    # ----- moyenne : encapsulée + validée + notification -----
    @property
    def moyenne(self) -> float:
        return self.__moyenne

    @moyenne.setter
    def moyenne(self, valeur: float) -> None:
        if not isinstance(valeur, (int, float)) or isinstance(valeur, bool):
            raise ValueError("La moyenne doit être un nombre.")
        if valeur < self.MOYENNE_MIN or valeur > self.MOYENNE_MAX:
            raise ValueError(
                f"La moyenne doit être comprise entre {self.MOYENNE_MIN} et "
                f"{self.MOYENNE_MAX} (reçu : {valeur})."
            )
        self.__moyenne = float(valeur)
        # Notification automatique des observateurs (Pattern Observer)
        self._notifier()

    # ----- liste_cours : accès contrôlé -----
    @property
    def liste_cours(self) -> List[Cours]:
        return list(self.__liste_cours)

    def ajouter_cours(self, cours: Cours) -> None:
        if not isinstance(cours, Cours):
            raise TypeError("ajouter_cours attend une instance de Cours.")
        self.__liste_cours.append(cours)

    # ----- polymorphisme TP3 -----
    def afficher_details(self) -> str:
        base = super().afficher_details()
        cours_txt = (
            ", ".join(c.nom_cours for c in self.__liste_cours)
            if self.__liste_cours
            else "aucun cours"
        )
        return (
            f"[Étudiant] {base} | N° {self.__numero_etudiant} | "
            f"Moyenne : {self.__moyenne:.2f}/20 | Cours : {cours_txt}"
        )

    def __repr__(self) -> str:
        return (
            f"Etudiant(nom={self.nom!r}, age={self.age}, "
            f"numero={self.__numero_etudiant!r}, moyenne={self.__moyenne})"
        )

File: java/src/main/java/com/ensta/gestion/Etudiant.java
package com.ensta.gestion;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Étudiant : Personne + numéro étudiant + moyenne + cours.
 *
 * Encapsulation TP2 : moyenne validée [0,20], numéro en lecture seule,
 * liste de cours protégée (copie défensive).
 *
 * Polymorphisme TP3 : afficherDetails() override.
 */
public class Etudiant extends Personne {

    public static final double MOYENNE_MIN = 0.0;
    public static final double MOYENNE_MAX = 20.0;

    private final String numeroEtudiant;   // lecture seule après création
    private double moyenne;
    private final List<Cours> listeCours;

    public Etudiant(String nom, int age, String numeroEtudiant, double moyenne) {
        super(nom, age);
        if (numeroEtudiant == null || numeroEtudiant.trim().isEmpty()) {
            throw new IllegalArgumentException("Le numéro d'étudiant ne peut pas être vide.");
        }
        this.numeroEtudiant = numeroEtudiant.trim();
        this.listeCours = new ArrayList<>();
        setMoyenne(moyenne);
    }

    public Etudiant(String nom, int age, String numeroEtudiant) {
        this(nom, age, numeroEtudiant, 0.0);
    }

    // ----- numero etudiant : lecture seule -----
    public String getNumeroEtudiant() {
        return numeroEtudiant;
    }

    // ----- moyenne -----
    public double getMoyenne() {
        return moyenne;
    }

    public void setMoyenne(double moyenne) {
        if (moyenne < MOYENNE_MIN || moyenne > MOYENNE_MAX) {
            throw new IllegalArgumentException(
                "La moyenne doit être comprise entre " + MOYENNE_MIN + " et "
                    + MOYENNE_MAX + " (reçu : " + moyenne + ")."
            );
        }
        this.moyenne = moyenne;
    }

    // ----- liste cours : accès contrôlé -----
    public List<Cours> getListeCours() {
        // Copie défensive : modifier la liste retournée n'altère pas l'interne.
        return Collections.unmodifiableList(new ArrayList<>(listeCours));
    }

    public void ajouterCours(Cours cours) {
        if (cours == null) {
            throw new IllegalArgumentException("Le cours ne peut pas être null.");
        }
        listeCours.add(cours);
    }

    // ----- polymorphisme TP3 -----
    @Override
    public String afficherDetails() {
        StringBuilder sb = new StringBuilder("[Étudiant] ")
            .append(super.afficherDetails())
            .append(" | N° ").append(numeroEtudiant)
            .append(" | Moyenne : ").append(String.format("%.2f", moyenne)).append("/20")
            .append(" | Cours : ");
        if (listeCours.isEmpty()) {
            sb.append("aucun cours");
        } else {
            for (int i = 0; i < listeCours.size(); i++) {
                if (i > 0) sb.append(", ");
                sb.append(listeCours.get(i).getNomCours());
            }
        }
        return sb.toString();
    }
}
